### PR TITLE
post [new]: created article monad in golang

### PR DESCRIPTION
**Pull request by mustafasegf**
post for golang in monad article

### CODE REVIEW

File: _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
...
+
+### Why don't use golang?
+
+Right now, golang jobs out there have increase a lot in devops space and also in web development space. Golagn is really simple, i bet it'll take you less than 1 week to get productive in it. Golang also use garbage collector and i don't need to think about borrow checker rule. **_But..._** golang don't have strong type system. It have null as a value, don't have algebraic data type and don't have option and result. This might look like i'm being petty but trust me. Debugging null pointer in golang is not fun if you have recursive data structure (taken from real life experience).
+
+### Can't we make it better?
+
+Before go 1.18, this problem have a hard solution. Since golang don't have generic, we couldn't make a generic container that could accept any type. We could use `interface{}` or reflection, but we'll loose the strict type and incur runtime performance penalty.
```
Comment (Ghibranalj) [inline _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md:17]: loose -> lose

File: _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+
+But fear no more. In golang 1.18, they introduce generic. Horray 🎉. Now we could make a simple container that could accept generic type.
+
+### let's add option and result to golang
+
+now, we could make a simple struct for the option like this
```
Comment (Ghibranalj) [inline _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md:23]: captialize N

File: _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+
+```go
+type Option[T any] struct {
+    value T
+    isSome bool
+}
+```
+
+and we could make one for result like this
```
Comment (Ghibranalj) [inline _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md:32]: add : maybe?

File: _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md [hunk 4]
```diff
@@ -0,0 +1,238 @@
...
+
+```go
+type Result[T any] struct {
+    value T
+    err error
+    isErr bool
+}
+```
+
+after that we just need to implement map, getter, unwrap, and constructor right? well... yeah but that's tedious and thankfully there's already a good golang package called [`samber/mo`](https://github.com/samber/mo) that already implement most of the stuff that we need.
```
Comment (Ghibranalj) [inline _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md:42]: capitalize A

File: _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md [hunk 5]
```diff
@@ -0,0 +1,238 @@
...
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusBadRequest)
+		return
+	}
+
+	w.Write([]byte("success"))
+}
+```
+
+this is a basic golang http server that do migration on start and server post request and add it to database. As you can see, golang need to check on the error, and if it's error then we need to early return it. If we don't check the error, the value of `user` will be `nil`.
```
Comment (Ghibranalj) [inline _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md:131]: Capitalize T

File: _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+		},
+		func(err error) (Option[User], error) {
+			http.Error(w, err.Error(), http.StatusInternalServerError)
+			return None[User](), err
+		},
+	)
+}
+```
+
+at a first glance, this look much more clutered. The reason why becasuse golang right now doesn't have light weight anonymous function declaration. There's an [ongoing proposal](https://github.com/golang/go/issues/21498) for it. But it looks like it won't be approved anytime soon.
```
Comment (Ghibranalj) [inline _posts/2023-06-23-adding_monad_to_golang-mustafasegf.md:230]: capitalize A

**Review by Ghibranalj — CHANGES_REQUESTED** [review]
Small mistakes at first skim

**Review by dnswd — APPROVED** [review]
my comments idem with @Ghibranalj but good overall

**Review by Ghibranalj — APPROVED** [review]

**Comment by Ghibranalj** [comment]
Looks good.
Im just not into the pic. But..... Whatever

**Comment by mustafasegf** [comment]
> Looks good. Im just not into the pic. But..... Whatever

I do agree, the image is subpar. I'll change it

**Pull request merged**